---
title: "E21.3 GO for WW"
author: "Chandler Sutherland"
date: "2025-08-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning=FALSE)
```

```{r}
library(ggplot2)
library(ggsignif)
library(ggpubr)
library(ggbeeswarm)
require(tidyverse)
library(openxlsx)
library(patchwork)
library(ggExtra)
library(scales)
library(topGO)
```

Copyright (c) Chandler Sutherland Email: [chandlersutherland\@berkeley.edu](mailto:chandlersutherland@berkeley.edu){.email}

Purpose: GO/Pfam enrichment analysis for WW promoter paper. 

Load WW's data and use her code to filter in the same way to pangenes > size 5 and mean middle leaf TPM > 1. 
```{r}
# Adapted WW's code for filtering 

#load files (changed to mine)
pangene_alltissue_mean=read.csv(file="data/pangene_alltissue_mean_toChandler.csv")
pangene_alltissue_adjCV=read.csv(file="data/pangene_alltissue_adjCV.csv")

#filter to pangenes >5 and TPM >= 1
pangene_alltissue_mean_filter=pangene_alltissue_mean %>% 
  subset(select=c("Pangene","Size","Similarity","middle")) %>% 
  filter(Size>5) %>% 
  filter(middle>=1)

#filter CV to these expressed and reasonably sized pangenes 
pangene_alltissue_adjCV_filter=pangene_alltissue_adjCV %>% 
  subset(select=c("Pangene","Size","Similarity","middle")) %>% 
  filter(Pangene %in% pangene_alltissue_mean_filter$Pangene) %>%
  mutate(middle=as.numeric(middle))

#plot as sanity check 
ggplot(data=pangene_alltissue_adjCV_filter, aes(x=as.numeric(middle))) +
  geom_density()+
  labs(x="Variability[log10(CV^2)]" )+
  xlim(-2,2)+
  theme_classic()

#scatter plot
pangene_CV_sim=merge(pangene_alltissue_adjCV[,c(1:3,5)],pangene_alltissue_mean[,c(1,5)],by="Pangene")
colnames(pangene_CV_sim)=c("Pangene","Size","Similarity","Adjusted_CV","Mean")
pangene_CV_sim$Adjusted_CV=as.numeric(pangene_CV_sim$Adjusted_CV)
quantile(pangene_CV_sim$Similarity,probs = c(0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1))
quantile(pangene_CV_sim$Adjusted_CV,probs = c(0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1),na.rm = TRUE)

#criteria for pangene selection for top % 
pangene_filter=pangene_CV_sim[pangene_CV_sim$Mean>=1&pangene_CV_sim$Size>5,]
set1=pangene_filter %>% filter(Similarity< 600 & Adjusted_CV>0.37) #20% and 20%
set2=pangene_filter %>% filter(Similarity< 667 & Adjusted_CV>0.3)  #25% and 25%
set3=pangene_filter %>% filter(Similarity< 746 & Adjusted_CV>0.23)  #30% and 30%

#plot sanity check 
ggplot() + 
  geom_point(data=pangene_filter,aes(x=log2(Similarity), y=Adjusted_CV),color="darkgrey",size=0.5, shape=20) +
  geom_point(data=set3,aes(x=log2(Similarity), y=Adjusted_CV),color="pink",size=0.5, shape=20) +
  geom_point(data=set2,aes(x=log2(Similarity), y=Adjusted_CV),color="red",size=0.5, shape=20) +
  geom_point(data=set1,aes(x=log2(Similarity), y=Adjusted_CV),color="darkred",size=0.5, shape=20) +
  #geom_smooth(data=pangene_filter,aes(x=log2(Similarity), y=Adjusted_CV),method=lm,linetype="solid", color="#AFD9FD",size=0.5,se=FALSE) +. #add a regression line
  geom_vline(xintercept = 9.54, linetype = "dashed", color = "pink") +
  geom_hline(yintercept = 0.23, linetype = "dashed", color = "pink") +
  geom_vline(xintercept = 9.38, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0.3, linetype = "dashed", color = "red") +
  geom_vline(xintercept = 9.22, linetype = "dashed", color = "darkred") +
  geom_hline(yintercept = 0.37, linetype = "dashed", color = "darkred") +
  xlab("log2(Similarity)") + ylab("Variability[log10(CV^2)]") + ylim(-1.5,1.5)+
  theme_classic()

#read in pangene membership info 
pan_gene <- read_tsv('data/pan-zea.v3.pan-genes_table_NAM.txt')
```

## GO Enrichment 

GO enrichment on: 
1. genes with highly variable transcription 
2. genes with both highly variable transcription AND promoter sequence 

Methods: 

The PANNZER2 annotations (Törönen et al. 2018) for the 26 NAM lines were downloaded from Maize GDB (https://download.maizegdb.org/GeneFunction_and_Expression/Pannzer_GO_Terms/B73_GO.out). Only annotations with a positive predictive value greater than 0.6 and an ARGOT rank of 1 were kept. All GO terms assigned to genes within a pangene were then transferred to their pangene. GO term enrichment analysis was then performed using TopGO (Alexa and Rahnenfuhrer 2023) version 2.36.0 and enrichment was calculated using the Fisher's exact test and the “elim” algorithm. Only GO terms that were assigned to 3 or more pangenes with highly variable expression and with enrichment P-values less than 0.05 were reported.

### Pannzer Annotation collation from Maize GDB 

"It is a parseable file, which uses a context sensitive grammar, as explained below. The file has six tab-separated columns labelled qpid, type, score, PPV, id and desc. 
- The first column (qpid) always contains the identifier of the query sequence. 
- The second column (type) can take the following values, and the score, PPV, id and desc columns change meaning accordingly:
    - original_DE: score is euk for eukaryotic query species and bac otherwise; 
    - id is the form factor of desc, the description from the input FASTA file 
    - qseq: desc is the amino acid sequence of the query protein 
    - DE: score is the prediction score; 
    - PPV is the normalized prediction score between 0 and 1; 
    - id is the form factor of desc, the predicted description 
    - GN: desc is the gene symbol. A gene symbol is predicted if its support is greater than half in the list of homologs ontology_predictor where ontology is one of MF (molecular function), BP (biological process), CC (cellular component) and predictor is one of RM3, ARGOT, HYGE or JAC. 
    - score is the prediction score, 
    - PPV is the normalized prediction score between 0 and 1, 
    - id is the GO identifier and desc is the short description of the GO class EC_predictor: 
    - desc is the GO class that has the highest PPV and has a link in ec2go, 
    - id is the EC class KEGG_predictor: 
    - desc is the GO class that has the highest PPV and has a link in kegg2go, 
    - id is the KEGG pathway identifier
    
```{r}
# I downloaded the maize GDB GO annotations from here: https://download.maizegdb.org/GeneFunction_and_Expression/Pannzer_GO_Terms/

#get paths 
annotations <- Sys.glob('//wsl.localhost//Ubuntu//home//chandlersutherland//e21_scratch//*_GO.out')

load_annotation <- function(path){
  nam <- path %>% str_split_i('//', 7) %>% str_split_i('_', 1)
  df <- read_tsv(path) %>% 
    filter(ARGOT_PPV > 0.6) %>%
    filter(ARGOT_rank==1) %>%
    mutate(goid=paste0('GO:', goid)) %>%
    mutate(nam=nam) %>%
    mutate(nam_ID=str_sub(qpid, 1, str_length(qpid) - 11))
  return(df)
}

pan_annotation <- do.call(rbind, lapply(annotations, load_annotation))
pan_annotation %>% group_by(nam, nam_ID, ontology) %>% summarize(n_annotations=n())
```

```{r}
#convert pangene list to long format 
nam_list <- colnames(pan_gene)[2:27]
long_pangene <- pan_gene %>% 
  pivot_longer(cols=nam_list, names_to='nam_id', values_to='qpid') %>%
  filter(qpid != 'NONE') %>%
  separate_longer_delim(qpid, delim=',') %>%
  mutate(qpid=str_replace_all(qpid, '_T', '_P'))

#add GO annotation 
pan_annotation_pangene <- merge(long_pangene, pan_annotation, by=c('qpid'))
pan_count <- pan_annotation_pangene %>% group_by(Pangene, ontology) %>% summarize(n_go=n_distinct(goid))

ggplot(pan_count)+
  geom_bar(aes(x=n_go), stat='count')+
  theme_classic()+
  labs(title='Number of GO terms assigned per Pangene', 
       xlab='Number distinct GO terms')+
  facet_wrap(~ontology)

pangene_annotation <- pan_annotation_pangene %>%
  subset(select=c(Pangene, ontology, goid, desc)) %>%
  unique()

#write out 
pangene_annotation %>% write.csv('output/pangene_GO_annotation.csv')

#convert to topGO list 
go_df <- aggregate(goid ~ Pangene, data=pangene_annotation, FUN=c)
pangene_go_list <- structure(go_df$goid, .Names=go_df$Pangene)
```

Multiple GO terms per ontology per gene is not atypical even in single genome analyses. Proceed. 30k pangenes with biological process annotations, 12k with CC annotations, and 33k with molecular function annotations. Yay!

### Highly variable expression
```{r}
#set limits of distribution and extremes
q <- quantile(pangene_alltissue_adjCV_filter$middle, probs=c(0.05, 0.95))

d <- density(pangene_alltissue_adjCV_filter$middle)
density_df <- data.frame(x=d$x, y=d$y) 

#pangene_alltissue_adjCV_filter %>%
distro <- ggplot(density_df, aes(x=x, y=y))+
  geom_line()+
  #geom_vline(xintercept=q[[1]], linetype=2, color='red')+
  #geom_vline(xintercept=q[[2]], linetype=2, color='red')+
  geom_ribbon(data=subset(density_df, x <= q[[1]]), 
              aes(ymin=0, ymax=y), fill='red', alpha=0.5)+
  geom_ribbon(data=subset(density_df, x >= q[[2]]), 
              aes(ymin=0, ymax=y), fill='red', alpha=0.5)+
  labs(x="Variability[log10(CV^2)]" )+
  xlim(-2, 2)+
  theme_classic()

n_top <- sum(pangene_alltissue_adjCV_filter$middle >= q[[2]])
n_bottom <- sum(pangene_alltissue_adjCV_filter$middle <= q[[1]])
distro
```

There are 759 pangenes in the top 5% and 759 in the bottom 5%.

Get the top 5% of CV values of middle: 

```{r}
#mark the top 5% 
q <- quantile(pangene_alltissue_adjCV_filter$middle, probs=c(0.05, 0.95))

#get a B73 list of top 5 or not 
top5_df <- pangene_alltissue_adjCV_filter %>% 
  mutate(top_5=case_when(middle>q[[2]] ~ 'yes', 
                         .default='no')) 
#transform to list format for topgo 
pangene_list <- factor(as.integer(top5_df$top_5=='yes'))
names(pangene_list) <- top5_df$Pangene
```

```{r}
run_topGO <- function(pangene_list, ontology, pangene_go_list){
  topGO_data <- new("topGOdata", 
                  ontology=ontology, 
                  allGenes=pangene_list, #yes or no gene list 
                  annot=annFUN.gene2GO, 
                  gene2GO=pangene_go_list)
  fishers_result <- runTest(topGO_data, algorithm='elim', statistic='fisher')

  fishers_table <- GenTable(topGO_data, Fishers=fishers_result, useLevels=T) %>%
    mutate(Fishers=as.numeric(Fishers)) %>%
    filter(Fishers < 0.05) %>% # p value > 0.05
    filter(Significant >= 3) %>% # at least 3 pangenes are within the GO Term 
    mutate(observed_expected_ratio=Significant/Expected)
  
  return(fishers_table)
}

plot_topGO <- function(fishers_table, ontology){
  p1 <- fishers_table %>%
    ggplot(aes(x=observed_expected_ratio, y=factor(reorder(Term, observed_expected_ratio))))+
    geom_point(aes(color=-log10(Fishers), size=Annotated))+
    scale_color_gradient(low='blue', high='red')+
    xlab("Observed count over expected count")+
    labs(color = "-Log10 p-value", 
         size = "Number of genes \nwith annotation", 
         title=paste('GO Analysis', ontology))+
    geom_rect(mapping=aes(xmin=-Inf, xmax=1, ymin=-Inf, ymax=Inf), color="grey", alpha=0.01) +
    ylab('')+
    theme_classic()
  return(p1)
}
  
p_bp <- run_topGO(pangene_list, 'BP', pangene_go_list) %>% plot_topGO('BP')
p_mf <- run_topGO(pangene_list, 'MF', pangene_go_list) %>% plot_topGO('MF')

p_bp / p_mf + plot_annotation(title='top 5%')    

#fishers_table %>% 
#  ggplot(aes(x=reorder(Term, -log10(Fishers)), y=-log10(Fishers), size=Significant, color=-log10(Fishers)))+
#  geom_point()+
#    scale_color_continuous(low = 'royalblue', high = 'tomato') +
#    xlab('') + ylab('-log10 (p value)') +
#    labs(title = paste0('GO Analysis (MF) Top 5% CV')) +
#    geom_hline(
#      yintercept = c(-log10(0.05), -log10(0.01), -log10(0.001)),
#      linetype = c("dotted", "dotted", "dotted"),
#      colour = c("black", "black", "black"),
#      size = c(1, 1, 1)
#    ) +
#    theme_bw(base_size = 12) +
#    coord_flip()
```
bottom 5%: 
```{r}
#get a B73 list of top 5 or not 
bottom5_df <- pangene_alltissue_adjCV_filter %>% 
  mutate(bottom_5=case_when(middle<q[[1]] ~ 'yes', 
                         .default='no')) 
#transform to list format for topgo 
pangene_list <- factor(as.integer(bottom5_df$bottom_5=='yes'))
names(pangene_list) <- bottom5_df$Pangene

p_bp_b <- run_topGO(pangene_list, 'BP', pangene_go_list) %>% plot_topGO('BP')
p_mf_b <- run_topGO(pangene_list, 'MF', pangene_go_list) %>% plot_topGO('MF')

p_bp_b / p_mf_b + plot_annotation(title='bottom 5%')   
```
```{r}
distro / (p_bp_b + p_bp)
```

### Genes with both highly variable transcription AND promoter sequence 
```{r}
scatter <- #plot sanity check 
ggplot() + 
  geom_point(data=pangene_filter,aes(x=log2(Similarity), y=Adjusted_CV),color="darkgrey",size=0.5, shape=20) +
  geom_point(data=set3,aes(x=log2(Similarity), y=Adjusted_CV),color="pink",size=0.5, shape=20) +
  geom_point(data=set2,aes(x=log2(Similarity), y=Adjusted_CV),color="red",size=0.5, shape=20) +
  geom_point(data=set1,aes(x=log2(Similarity), y=Adjusted_CV),color="darkred",size=0.5, shape=20) +
  #geom_smooth(data=pangene_filter,aes(x=log2(Similarity), y=Adjusted_CV),method=lm,linetype="solid", color="#AFD9FD",size=0.5,se=FALSE) +. #add a regression line
  geom_vline(xintercept = 9.54, linetype = "dashed", color = "pink") +
  geom_hline(yintercept = 0.23, linetype = "dashed", color = "pink") +
  geom_vline(xintercept = 9.38, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 0.3, linetype = "dashed", color = "red") +
  geom_vline(xintercept = 9.22, linetype = "dashed", color = "darkred") +
  geom_hline(yintercept = 0.37, linetype = "dashed", color = "darkred") +
  xlab("log2(Similarity)") + ylab("Variability[log10(CV^2)]") + ylim(-1.5,1.5)+
  theme_classic()
pangene_filter=pangene_CV_sim[pangene_CV_sim$Mean>=1&pangene_CV_sim$Size>5,]

set1 <- pangene_alltissue_adjCV_filter %>% 
  mutate(set_1=case_when(Similarity<600 & middle>0.37 ~ 'yes', 
                         .default='no')) 
#transform to list format for topgo 
pangene_list <- factor(as.integer(set1$set_1=='yes'))
names(pangene_list) <- set1$Pangene

p_bp_set1 <- run_topGO(pangene_list, 'BP', pangene_go_list) %>% plot_topGO('BP')
p_mf_set1 <- run_topGO(pangene_list, 'MF', pangene_go_list) %>% plot_topGO('MF')

p_bp_set1 / p_mf_set1 + plot_annotation(title='Set 1: top 20% CV and promoter variability')   
```

